miner/src/pool/replace.rs

```diff
@@ -0,0 +1,416 @@
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Replacing Transactions
+//!
+//! When queue limits are reached, a new transaction may replace one already
+//! in the pool. The decision whether to reject, replace or retain both is
+//! delegated to an implementation of `ShouldReplace`.
+//!
+//! Here we decide based on the sender, the nonce and gas price, and finally
+//! on the `Readiness` of the transactions when comparing them
+
+use std::cmp;
+
+use ethereum_types::{U256, H160 as Address};
+use txpool::{self, scoring::{Choice, Scoring}, ReplaceTransaction};
+use txpool::VerifiedTransaction;
+use super::{client, ScoredTransaction};
+
+/// Choose whether to replace based on the sender, the score and finally the
+/// `Readiness` of the transactions being compared.
+#[derive(Debug)]
+pub struct ReplaceByScoreAndReadiness<S, C> {
+	scoring: S,
+	client: C,
+}
+
+impl<S, C> ReplaceByScoreAndReadiness<S, C> {
+	/// Create a new `ReplaceByScoreAndReadiness`
+	pub fn new(scoring: S, client: C) -> Self {
+		ReplaceByScoreAndReadiness { scoring, client }
+	}
+}
+
+impl<T, S, C> txpool::ShouldReplace<T> for ReplaceByScoreAndReadiness<S, C>
+where
+	T: VerifiedTransaction<Sender = Address> + ScoredTransaction + PartialEq,
+	S: Scoring<T>,
+	C: client::NonceClient,
+{
+	fn should_replace(
+		&self,
+		old: &ReplaceTransaction<T>,
+		new: &ReplaceTransaction<T>,
+	) -> Choice {
+		let both_local = old.priority().is_local() && new.priority().is_local();
+		if old.sender() == new.sender() {
+			// prefer earliest transaction
+			match new.nonce().cmp(&old.nonce()) {
+				cmp::Ordering::Equal => self.scoring.choose(&old, &new),
+				_ if both_local => Choice::InsertNew,
+				cmp::Ordering::Less => Choice::ReplaceOld,
+				cmp::Ordering::Greater => Choice::RejectNew,
+			}
+		} else if both_local {
+			Choice::InsertNew
+		} else {
+			let old_score = (old.priority(), old.gas_price());
+			let new_score = (new.priority(), new.gas_price());
+			if new_score > old_score {
+				let state = &self.client;
+				// calculate readiness based on state nonce + pooled txs from same sender
+				let is_ready = |replace: &ReplaceTransaction<T>| {
+					let state_nonce = state.account_nonce(replace.sender());
+					let nonce =
+						replace.pooled_by_sender.map_or(state_nonce, |txs| {
+							txs.iter().fold(state_nonce, |nonce, tx| {
```

Comment:
I think we should early-exit here on first unmatching nonce. Also regular `for` loop might be more readable in such case.